util/git/client.go

```diff
@@ -642,16 +654,65 @@ func (m *nativeGitClient) lsRemote(revision string) (string, error) {
 			return hash, nil
 		}
 	}
+
 	// We support the ability to use a truncated commit-SHA (e.g. first 7 characters of a SHA)
 	if IsTruncatedCommitSHA(revision) {
 		log.Debugf("revision '%s' assumed to be commit sha", revision)
 		return revision, nil
 	}
+
 	// If we get here, revision string had non hexadecimal characters (indicating its a branch, tag,
 	// or symbolic ref) and we were unable to resolve it to a commit SHA.
 	return "", fmt.Errorf("Unable to resolve '%s' to a commit SHA", revision)
 }
 
+func (m *nativeGitClient) resolveSemverRevision(revision string, refs []*plumbing.Reference) (string, bool) {
+	constraint, err := semver.NewConstraint(revision)
+	if err != nil {
+		return "", false
+	}
+
+	versions := make(map[string]*semver.Version)
+	for _, ref := range refs {
+		if !ref.Name().IsTag() {
+			continue
+		}
+
+		tag := ref.Name().Short()
+		version, err := semver.NewVersion(tag)
+		if err != nil {
+			if errors.Is(err, semver.ErrInvalidSemVer) {
+				log.Debugf("Invalid semantic version: %s", tag)
+				continue
+			}
+
+			log.Debugf("invalid constraint in tags: %s", err.Error())
+
+			return "", false
+		}
+
+		if constraint.Check(version) {
+			versions[ref.Hash().String()] = version
+		}
+	}
+
+	// Only if versions has items we're going to iterate, otherwise it'll follow the normal flow.
+	if len(versions) != 0 {
+		maxVersion := semver.New(0, 0, 0, "", "")
+		maxVersionHash := plumbing.ZeroHash.String()
+		for hash, version := range versions {
+			if version.GreaterThan(maxVersion) {
+				maxVersion = version
+				maxVersionHash = hash
+			}
+		}
+
+		return maxVersionHash, true
+	}
```

Comment:
I think we can combine this with the loop above while computing the versions itself. Or rather, instead of storing the map, maybe we can directly compute the `maxVersion` and `maxVersionHash` itself. WDYT?

Reply:
I think `lsRemote` method will be a little clunky trying to add that into the loop! Adding all the logic there will make it hard to read, I'm saying this 'cause I tried to glue all together!

Also, if your comment is because you're worried about the performance I'd say there's just one case where the two loops will run: When the provided tag (which matches the semver pattern) does not exist!
Other than that the loops will run only once because if the provided revision is:
* __v0.1.*__, will run only the semantic version loop
* __v0.1.2__, will run only the semantic version loop
* __custom-tag__, will run only the default (old?) loop because "custom-tag" does not match semantic versioning
* __master__, will run only the default (old?) loop because "master" does not match semantic versioning

So I think will be good to have them separated! WDYT @ishitasequeira??

Reply:
@ishitasequeira did you have a chance to see here?

Reply:
Thanks for the explanation!! I am good with the way it is. 

It would be good to add the above example in comments for easier understanding for folks looking at the code later. The explanation can be just above the function itself.

Reply:
I've just pushed it @ishitasequeira!